src/averageAges.js

```diff
@@ -55,7 +76,29 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothers = people.filter(
+    (person) => people.some((child) => person.name === child.mother)
+  );
+  const children = people.filter(
+    (person) => people.some((mother) => person.mother === mother.name)
+  );
+  const mothersWithSon = mothers.filter(
+    (person) => children.some(
+      (child) => child.sex === 'm' && child.mother === person.name)
+  );
+  const sons = children.filter(
+    (person) => person.sex === 'm');
+  const avarageDifference = (onlyWithSon)
+    ? sons.reduce(
+      (acc, son) => acc + son.born
+      - mothersWithSon.find((mother) => son.mother === mother.name).born, 0)
+      / sons.length
+    : children.reduce(
```

Comment:
Don't repeat yourself. Try to find a solution where you need only one `reduce` method.
Hint - you can create an array of `children` and use it with reduce. This array should depend on `onlyWithSon` argument. You can use a ternary operator for this purpose (like I've shown in examples above)

Reply:
Thank you for such detailed explanations.